lib/compress/zstd_double_fast.c

```diff
@@ -11,8 +11,43 @@
 #include "zstd_compress_internal.h"
 #include "zstd_double_fast.h"
 
+static void ZSTD_fillDoubleHashTableForCDict(ZSTD_matchState_t* ms,
+                              void const* end, ZSTD_dictTableLoadMethod_e dtlm)
+{
+    const ZSTD_compressionParameters* const cParams = &ms->cParams;
+    U32* const hashLarge = ms->hashTable;
+    U32  const hBitsL = cParams->hashLog;
+    U32  const mls = cParams->minMatch;
+    U32* const hashSmall = ms->chainTable;
+    U32  const hBitsS = cParams->chainLog;
+    const BYTE* const base = ms->window.base;
+    const BYTE* ip = base + ms->nextToUpdate;
+    const BYTE* const iend = MIN((const BYTE*)end, base + (1u << (32 - ZSTD_SHORT_CACHE_TAG_BITS))) - HASH_READ_SIZE;
```

Comment:
I'm not sure I understand. You're limiting the dict fill to the first 16MB of the dict right? But for very large dictionaries we'd rather limit it to the last 16MB.

Well, actually for very large dictionaries we'd probably rather not do short cache. But I guess that's not an option as implemented.

You might want to hook into the logic here?:

https://github.com/facebook/zstd/blob/v1.5.2/lib/compress/zstd_compress.c#L4224-L4229

Reply:
Good point, will change so it uses the last 16MB.

Reply:
Fast hash tables will not effectively index more than 16MB. You'd need a O(16M) entry hash table. If someone wanted to index a larger table, they'd need to either use LDM, or a higher compression level.

But yeah, we definitely want the last 16MB.

Reply:
@terrelln @felixhandte I discussed offline with Felix and he raised concerns about non-standard usecases. E.g. seekable compression where every 10th frame is used as a dictionary for the other 9 frames. You could imagine someone creating a 16MB CDict for this. 

Nick, you're saying that because fast/dfast hashtables are < 1MB, they can't store 16MB worth of positions, so we don't need to worry. Is my understanding correct?

Reply:
Yes

The fast/dfast match finders will never be good at finding matches > 16 MB in the past. So there is no reason to support it IMO.

We have a great match finder for that purpose, LDM, and if someone wants that behavior they should use that.